strings/base_collections_base.h

```diff
@@ -70,43 +140,66 @@ WINRT_EXPORT namespace winrt
 
             T Current() const
             {
-                if (m_current == m_end)
+                return m_owner->perform_shared([&]
                 {
-                    throw hresult_out_of_bounds();
-                }
+                    this->check_version(*m_owner);
 
-                if constexpr (!impl::is_key_value_pair<T>::value)
-                {
-                    return m_owner->unwrap_value(*m_current);
-                }
-                else
-                {
-                    return make<impl::key_value_pair<T>>(m_owner->unwrap_value(m_current->first), m_owner->unwrap_value(m_current->second));
-                }
+                    if (m_current == m_end)
+                    {
+                        throw hresult_out_of_bounds();
+                    }
+
+                    return current_value();
+                });
             }
 
-            bool HasCurrent() const noexcept
+            bool HasCurrent() const
             {
-                return m_current != m_end;
+                return m_owner->perform_shared([&]
+                {
+                    this->check_version(*m_owner);
+                    return m_current != m_end;
+                });
             }
 
-            bool MoveNext() noexcept
+            bool MoveNext()
             {
-                if (m_current != m_end)
+                return m_owner->perform_exclusive([&]
                 {
-                    ++m_current;
-                }
+                    this->check_version(*m_owner);
```

Comment:
Why is exclusive required for iterator movement? If you acquire the shared lock and verify the version info, then m_end should still be accurate, and this otherwise does not mutate the target collection.

Reply:
The iterator increment (`++m_current`) needs to be done under *some* exclusive lock, because the underlying container (e.g. `std::vector`) iterators provide no guarantees of atomicity.

My guess is this is a compromise to avoid adding a lock object to every iterator instance - when an iterator needs to lock itself exclusively, it simply piggy-backs on its owner's exclusive lock. This also means we only need to take a single lock, instead of taking the iterator's lock exclusive and the owner's lock shared (or would it be in the other order?). Doing a dance with two locks raises the complexity and makes it more likely that we'll accidentally introduce a deadlock.

But as a result, we're taking an exclusive lock, blocking other benign concurrent operations from running.

So I think I get the rationale, but I'm not completely, totally, 100% convinced it's the right decision. More like 80%.

Reply:
Ah, I see. vector::iterator::operator++ might "do something" to the underlying vector (like, say, in debug mode?) so we have to go exclusive here as we can't assume anything about it.

Reply:
Ryan's pretty much summarized my thoughts. From the PR description:

> The most noteworthy "flaw" is that mutating iterator operations lock the collection exclusively. This is mostly a limitation imposed by the base IIterable type which uses opaque iterators for the IIterator objects. In theory, the collection only needs to be locked in shared mode, however that would require the IIterator objects themselves to have a separate lock, which seemed like overkill and unlikely to be worth it. Compare this to the existing ABI collection types which know the underlying vector/map types and can perform interlocked operations - in the case of vector - or forgo sequential consistency while still guaranteeing safety - in the case of map.

Personally I don't think it's worth the complexity and the breaking change associated with "piping" information down to the base iterable type, but I'm of course open to other opinions.

Deadlock is always something to look out with when multiple locks are involved, however I'm fairly certain that acquiring the iterator lock first and then the iterable's lock second is guaranteed to be correct.